File: gui/linkers/modules/utils/urlChecker.js
import { lazySplit, isOnlyDigits } from "./utils.js"
import { firstURLLazySplit } from "./websiteUtils.js"

function checkUrl(url, urlInfo){
    let [index, tag] = firstURLLazySplit(url, 0)

    let possibleTags = []
    let failedAttemps = 0
    let tagsConcatenation = []

    while(index !== -1 && failedAttemps<urlInfo.maxFailedAttemps){
        [index, tag] = lazySplit(url, index)

        if(!tag || tag[0] === "?" || urlInfo.invalidCategories.has(tag) || isOnlyDigits(tag) || tag.length>20) failedAttemps += 1
        else{
            tagsConcatenation.push(tag)
            possibleTags.push(tagsConcatenation.join("/"))
        }
    }

    return possibleTags
}

function sequentialURLCheck(urls, urlInfo){
    let s = new Set()

    for(let i=urls.length-1; i>=0; i--){
        let possibleTags = checkUrl(urls[i], urlInfo)
        for(let j=possibleTags.length-1; j>=0; j--) s.add(possibleTags[j])
    }
    
    return [...s]
}

export { checkUrl, sequentialURLCheck }

File: engine/News/urlChecker.py
from multiprocessing.pool import Pool
from Utils.utils import lazySplit, isOnlyDigits, firstURLLazySplit
from Utils.dataTypes import URLInfo
from itertools import chain
import time

def init(urlInfo):
    global urlInfoArg
    urlInfoArg = urlInfo

def checkUrlParallely(url):
    return checkUrl(url, urlInfoArg)

def checkUrl(url, urlInfo=None):
    if(urlInfo is None): return None

    index, tag = firstURLLazySplit(url, 0)
    possibleTags = []
    failedAttemps = 0
    tagsConcatenation = ""

    while(index != -1 and failedAttemps<urlInfo.maxFailedAttemps):
        index, tag = lazySplit(url, index)
        if(tag == "" or isOnlyDigits(tag) or tag in urlInfo.invalidTags):
            failedAttemps += 1
        else:
            tagsConcatenation = "".join([tagsConcatenation, "/", tag]) if tagsConcatenation != "" else tag

            if(tagsConcatenation not in urlInfo.invalidTags):
                possibleTags.append(tagsConcatenation)

    return possibleTags

def parallelURLCheck(urls, urlInfo, numPools):
    with Pool(numPools, init, (urlInfo,)) as pool:
        return set(chain.from_iterable(pool.map(checkUrlParallely, urls)))

def sequentialURLCheck(urls, urlInfo):
    s = set()

    for url in urls:
        s.update(checkUrl(url, urlInfo))
    
    return s
